builtins/web/blob.cpp

```diff
@@ -0,0 +1,488 @@
+  if (self == proto_obj) {
+    return api::throw_error(cx, api::Errors::WrongReceiver, "size get", "Blob");
+  }
+
+  auto blob = Blob::blob(self);
+  args.rval().setNumber(blob->size());
+  return true;
+}
+
+bool Blob::type_get(JSContext *cx, unsigned argc, JS::Value *vp) {
+  METHOD_HEADER(0);
+  // TODO: Change this class so that its prototype isn't an instance of the class
+  if (self == proto_obj) {
+    return api::throw_error(cx, api::Errors::WrongReceiver, "type get", "Blob");
+  }
+
+  auto type = Blob::type(self);
+  args.rval().setString(type);
+  return true;
+}
+
+std::vector<uint8_t> *Blob::blob(JSObject *self) {
+  auto blob = static_cast<std::vector<uint8_t> *>(
+      JS::GetReservedSlot(self, static_cast<size_t>(Blob::Slots::Data)).toPrivate());
+
+  MOZ_ASSERT(blob);
+  return blob;
+}
+
+JSString *Blob::type(JSObject *self) {
+  auto type = static_cast<JSString *>(
+      JS::GetReservedSlot(self, static_cast<size_t>(Blob::Slots::Type)).toPrivate());
+
+  MOZ_ASSERT(type);
+  return type;
+}
+
+bool Blob::append_value(JSContext *cx, HandleObject self, HandleValue val) {
+  auto blob = Blob::blob(self);
+
+  if (val.isObject()) {
+    RootedObject obj(cx, &val.toObject());
+
+    if (Blob::is_instance(obj)) {
+      auto src = Blob::blob(obj);
+      blob->insert(blob->end(), src->begin(), src->end());
+      return true;
+    } else if (JS_IsArrayBufferViewObject(obj) || JS::IsArrayBufferObject(obj)) {
+      auto span = value_to_buffer(cx, val, "Blob Parts");
+      if (span.has_value()) {
+        blob->insert(blob->end(), span->begin(), span->end());
+      }
+      return true;
+    }
+  } else if (val.isString()) {
+    auto chars = core::encode(cx, val);
+    if (!chars) {
+      return false;
+    }
+
+    // TODO: convert line endings: https://w3c.github.io/FileAPI/#convert-line-endings-to-native
+    auto src = chars.ptr.get();
+    auto len = chars.len;
+    blob->insert(blob->end(), src, src + len);
+    return true;
+  }
+
+  // FALLBACK: if we ever get here convert, to string and call append again
+  auto str = JS::ToString(cx, val);
+  if (!str) {
+    return false;
+  }
+
+  RootedValue str_val(cx, JS::StringValue(str));
+  return append_value(cx, self, str_val);
+}
+
+bool Blob::init_blob_parts(JSContext *cx, HandleObject self, HandleValue value) {
+  JS::ForOfIterator it(cx);
+  if (!it.init(value, JS::ForOfIterator::AllowNonIterable)) {
+    return false;
+  }
+
+  bool is_typed_array = value.isObject() && JS_IsTypedArrayObject(&value.toObject());
+  bool is_iterable = value.isObject() && it.valueIsIterable();
+
+  if (is_typed_array) {
+    // append typed array value directly...
+    return append_value(cx, self, value);
+   } else if (is_iterable) {
+    // if the object is an iterable, walk over its elements...
+    JS::Rooted<JS::Value> item(cx);
+    while (true) {
+      bool done;
+
+      if (!it.next(&item, &done)) {
```

Comment:
The iterator fails here for code like this:

```js
  var obj = {
    [Symbol.iterator]: Array.prototype[Symbol.iterator],
    get length() { throw test_error; }
  };
  assert_throws_exactly(test_error, function() {
    new Blob(obj);
  });
```
I'm not sure if this require a different approach or perhaps fixing in SpiderMonkey.

Reply:
Currently, the Blob constructor in this code throws a `TypeError`, but the test expects the length `getter` to be invoked, which should throw `test_error`:

```
Threw object "TypeError: Blob.constructor: blobParts must be an object" but expected it to throw object "[object Object]"
```

The issue is that the length function is never called because the iterator fails on the `it.next()` function.

Reply:
Ah, I see. I'll make some suggested changes to the code that I think should resolve this, though some tweaks might be required.

Reply:
I think those two suggestions should make this work. Essentially, ensure that we only throw the "must be an object" error when that is actually the issue, and otherwise keep whatever pending exception we have active.

Reply:
Oh nice, yeah this fixed few failing tests, thanks!